Reduce API list_volume when Attaching and Mounting Volumes

sarum90 opened the pull request:
With this PR these IStateChanges are constructed with blockdevice_ids. This means that they do not have to do the dataset_id -> blockdevice_id mapping as part of their execution. This enables the removal of list_volumes() calls at the beginning of the execution of either of them.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/clusterhq/flocker/2065)
<!-- Reviewable:end -->

exarkun commented:
Thanks!  Looks pretty good, just a few minor points.  One other thing that would be interesting is to see a count of the number of API calls made by a run of the acceptance test suite with and without this change.  I think our logs have this information so it's just a matter of a couple greps.  If you could do that and post the results somewhere, I think that'd be cool.  Then, please address the other review comments and merge.  Thanks!
<!-- Reviewable comment -K1K2DHhBveuyuD1qmEx:882977313 -->

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[flocker/node/agents/blockdevice.py, line 450 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1K-mI9WVdjGzcRZb3W)** ([raw file](https://github.com/clusterhq/flocker/blob/e54e04f0b021e7b3e4cd677c4dca7bc872291052/flocker/node/agents/blockdevice.py#L450)):</sup>
Pre-existing bad-naming here, sorry about that.  This is really the identifier of the `IBlockDeviceAPI`-managed volume.
<!-- Reviewable comment -K1K-mI9WVdjGzcRZb3X:1997508728 -->

---

<sup>**[flocker/node/agents/blockdevice.py, line 554 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1K06MUb9JuaTNYElAY)** ([raw file](https://github.com/clusterhq/flocker/blob/e54e04f0b021e7b3e4cd677c4dca7bc872291052/flocker/node/agents/blockdevice.py#L554)):</sup>
Likewise here.
<!-- Reviewable comment -K1K06MUb9JuaTNYElAZ:-1123490612 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 1203 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1Jzw3xhN5O76QKmbHr)** ([raw file](https://github.com/clusterhq/flocker/blob/e54e04f0b021e7b3e4cd677c4dca7bc872291052/flocker/node/agents/test/test_blockdevice.py#L1203)):</sup>
It might be nice (ie, do this if you feel like it, not required) to put this logic into a helper function for use throughout this test module.  
<!-- Reviewable comment -K1Jzw3y7jvjzbwylNbl:1779845731 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 3549 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1K0nx2_XJIbhdOPrsY)** ([raw file](https://github.com/clusterhq/flocker/blob/7a2623564155191b41937f0a3544224b69362aa4/flocker/node/agents/test/test_blockdevice.py#L3549)):</sup>
I wonder if it's worth replacing this with a test for the new failure behavior we'll get.  The failure from `destroy_volume`, `UnknownVolume`, will now be passed through.  An argument against doing this is that we probably don't much care exactly how this fails (demonstrated by the fact that I don't care that this PR changes the exception from `DatasetWithoutVolume` to `UnknownVolume`).

The value might be in proving that `AttachVolume.run` returns a `Deferred` that's connected to the underlying `IBlockDeviceAPI` call at all, though.  If the implementation returned `None` then the system wouldn't wait for the operation to complete, which is probably bad, and I don't think any other test would catch this.  Likewise, if it just returned `succeed(None)` then things wouldn't wait (and any test that just wanted a Deferred would still be satisfied).

So... yea... maybe there is value in having this negative-case test.
<!-- Reviewable comment -K1K0nx2_XJIbhdOPrsZ:559501531 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clusterhq/flocker/2065)
<!-- Sent from Reviewable.io -->

sarum90 commented:
Wrote a tiny python blob to count for me, ran on remote-logs from flocker/acceptance/aws/centos-7/aws:
[process.py.txt](https://github.com/ClusterHQ/flocker/files/19598/process.py.txt)

./both.txt 1661  - Includes This PR and #2057
./first.txt 1688    - only includes #2057
./before.txt 1780 - without # 2057 or this PR

So looks like the big wins are already in master (#2057), but this does drop order 1% of calls.

Now going to look through the review, address and merge.

sarum90 commented:
Re-running buildbot runs and merging once I've convinced them to be green.
<!-- Reviewable comment -K1KuDxP6vJ6E_QpFYmO:520579681 -->

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

<sup>**[flocker/node/agents/blockdevice.py, line 450 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1K-mI9WVdjGzcRZb3W)** ([raw file](https://github.com/clusterhq/flocker/blob/e54e04f0b021e7b3e4cd677c4dca7bc872291052/flocker/node/agents/blockdevice.py#L450)):</sup>
Done. Changed all docstrings in the file (Even some from my earlier PR).
<!-- Reviewable comment -K1Ko850gGxW48OsiaPC:-1651794649 -->

---

<sup>**[flocker/node/agents/blockdevice.py, line 554 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1K06MUb9JuaTNYElAY)** ([raw file](https://github.com/clusterhq/flocker/blob/e54e04f0b021e7b3e4cd677c4dca7bc872291052/flocker/node/agents/blockdevice.py#L554)):</sup>
Done.
<!-- Reviewable comment -K1KoASusH05PIqd-0bx:-499146009 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 1203 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1Jzw3xhN5O76QKmbHr)** ([raw file](https://github.com/clusterhq/flocker/blob/e54e04f0b021e7b3e4cd677c4dca7bc872291052/flocker/node/agents/test/test_blockdevice.py#L1203)):</sup>
Done.
<!-- Reviewable comment -K1KpqybownEpSuinYzR:-499146009 -->

---

<sup>**[flocker/node/agents/test/test_blockdevice.py, line 3549 \[r1\]](https://reviewable.io:443/reviews/clusterhq/flocker/2065#-K1K0nx2_XJIbhdOPrsY)** ([raw file](https://github.com/clusterhq/flocker/blob/7a2623564155191b41937f0a3544224b69362aa4/flocker/node/agents/test/test_blockdevice.py#L3549)):</sup>
Added it back, made it work with UnknownVolume. hadn't thought about your second point.
<!-- Reviewable comment -K1KtXwJJX_CqlRlgCjC:825785201 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clusterhq/flocker/2065)
<!-- Sent from Reviewable.io -->
